pin/set.go

```diff
@@ -0,0 +1,327 @@
+package pin
+
+import (
+	"bytes"
+	"crypto/rand"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"hash/fnv"
+	"io"
+	"sort"
+	"unsafe"
+
+	"github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/gogo/protobuf/proto"
+	"github.com/ipfs/go-ipfs/merkledag"
+	"github.com/ipfs/go-ipfs/pin/internal/pb"
+	"github.com/ipfs/go-ipfs/util"
+	"golang.org/x/net/context"
+)
+
+const (
+	defaultFanout = 256
+	maxItems      = 8192
+)
+
+func randomSeed() (uint32, error) {
+	var buf [4]byte
+	if _, err := rand.Read(buf[:]); err != nil {
+		return 0, err
+	}
+	return binary.LittleEndian.Uint32(buf[:]), nil
+}
+
+func hash(seed uint32, k util.Key) uint32 {
+	var buf [4]byte
+	binary.LittleEndian.PutUint32(buf[:], seed)
+	h := fnv.New32a()
+	_, _ = h.Write(buf[:])
+	_, _ = io.WriteString(h, string(k))
+	return h.Sum32()
+}
+
+type itemIterator func() (k util.Key, data []byte, ok bool)
+
+type keyObserver func(util.Key)
+
+type refcount uint8
```

Comment:
Oh and the three `refcount(X)` instances would just end up replaced by `byte(X)`; they're there to make a constant typed the right way, or to convert from uint64 to a refcount. The other way there's one `uint64(buf[idx])` that actually is missing the right logic to adapt to type refcount changes -- that's a bug.